### ISSUE TITLE
Automatically Estimated Minimum Number of Bootstrap Samples

### ISSUE THREAD

**Issue by alexsherstinsky** [scipy/scipy#14306]
**Is your feature request related to a problem? Please describe.**
The feature request will add a valuable capability to the "https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.bootstrap.html" algorithm.  Specifically, it will allow an up to two orders of magnitude reduction in the number of bootstrap samples (e.g., from the default of 10,000 down to 100).  Equally importantly, this feature will also determine that a larger than than default number of bootstrap samples is needed, if the data sample requires it to be so.

**Describe the solution you'd like**
The solution would implement an algorithm for choosing the optimal number of bootstrap samples, following the approach presented in "http://dido.econ.yale.edu/~dwka/pub/p1001.pdf", published in the year 2000 -- the full citation appears below:
```
@article{Andrews2000a,
  added-at = {2008-04-25T10:38:44.000+0200},
  author = {Andrews, Donald W. K. and Buchinsky, Moshe},
  biburl = {https://www.bibsonomy.org/bibtex/28e2f0a58cdb95e39659921f989a17bdd/smicha},
  day = 01,
  interhash = {778746398daa9ba63bdd95391f1efd37},
  intrahash = {8e2f0a58cdb95e39659921f989a17bdd},
  journal = {Econometrica},
  keywords = {imported},
  month = Jan,
  note = {doi: 10.1111/1468-0262.00092},
  number = 1,
  pages = {23--51},
  timestamp = {2008-04-25T10:38:52.000+0200},
  title = {A Three-step Method for Choosing the Number of Bootstrap Repetitions},
  url = {http://www.blackwell-synergy.com/doi/abs/10.1111/1468-0262.00092},
  volume = 68,
  year = 2000
}
```
This algorithm works as follows (as an iterative loop):
a) estimate the number of bootstrap samples;
b) generate these samples;
c) compute statistics on this `ndarray`, and use them to update the estimate of the number of bootstrap samples from step (a);
(d) repeat these steps till convergence. 

Because the optimal number of bootstrap samples is taken as the maximum of the estimates up to the current iteration, the procedure converges very rapidly (often in just 3 -- 5 iterations).

The initial contribution would focus on optimizing the standard error estimate of the bootstrap samples.  The next effort would be devoted to adding the support for confidence intervals.  After that, applicability to other statistics an be considered.

**Describe alternatives you've considered**
* Using the default number of bootstrap samples (which is 9,999 according to "https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.bootstrap.html").
* Guessing the number of bootstrap samples based on preliminary size and analysis of the data set.

**Additional context (e.g. screenshots)**
There is an initial implementation (please see https://github.com/great-expectations/great_expectations/tree/feature/GE-160/GE-237/alexsherstinsky/optimal_num_samples_bootstrapped_range_parameter_builder-2021_06_23-37/great_expectations/rule_based_profiler/estimators), which is focused on optimizing the standard error estimate of the bootstrap samples.  The first task would be to determine whether or not it can be adapted to the programming style of "https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.bootstrap.html" -- and, if affirmative, issue a pull request.

**Comment by tupui** [comment]
Thanks for the proposal @alexsherstinsky! This sounds like a good idea that look simple to implement. Do you propose to do a PR?

**Comment by alexsherstinsky** [comment]
> Thanks for the proposal @alexsherstinsky! This sounds like a good idea that look simple to implement. Do you propose to do a PR?

@tupui Thank you for getting back to me -- I would love to do a PR!  I would also like to confirm with @mdhaber that I can proceed according to the plan I outlined in the feature request (since he has been helping me tremendously), or if I should alter the course of action in any way.  Thank you very much.

**Comment by mdhaber** [comment]
One question is how the user will specify which they are interested in (standard error, CI, or some new output like bias) as the number of bootstrap samples will depend on that. Originally, I suggested `n_resamples=None` in the sense of "no fixed number", but that would not distinguish between SE and CI. We could allow a string like `n_resamples='auto-ci'`, and i would probably prefer that over a separate argument, but I'm not sure what's best.

I'll suggest a way to tie this feature into the existing code shortly.

**Comment by alexsherstinsky** [comment]
> One question is how the user will specify which they are interested in (standard error, CI, or some new output like bias) as the number of bootstrap samples will depend on that. Originally, I suggested `n_resamples=None` in the sense of "no fixed number", but that would not distinguish between SE and CI. We could allow a string like `n_resamples='auto-ci'`, and i would probably prefer that over a separate argument, but I'm not sure what's best.
> 
> I'll suggest a way to tie this feature into the existing code shortly.

@mdhaber I agree that we need to specify two aspects of the configuration somehow:
* Which mode do we want: SE?  CI?
* Do we want to use `auto` or not?

Using your example, if we could do it as:
* n_resamples='auto-se'; and
* n_resamples='auto-ci',
respectively.

I feel that this could be a bit confusing for the user, who today can simply obtain `result.standard_error` and `result.confidence_intervals` without any additional configuration.  With the above approach, the user could configure the call to `bootstrap()` with `n_resamples='auto-ci'`, but then request `result.standard_error`, which would not be applicable.

Would it make sense for us to try the following instead:
* Configure the automatic method for determining the number of bootstrap samples using `n_resamples='auto';
* If `n_resamples` == `auto`, then compute internally the automatic number of bootstrap samples for all supported use cases (SE and CI; maybe others later).
* Support the existing `result.standard_error` and `result.confidence_intervals` output accessors without any API changes.

The advantage of this is that we can release the support for automatically computing the number of bootstrap samples one more at a time (e.g., the SE mode first; the CI mode after that) with meaningful release comments offered to the users.

Looking forward to your thoughts as well as to the guidance for how to work this feature into the existing code.  Thank you!

**Comment by mdhaber** [comment]
Yeah, I thought about that ("compute internally the automatic number of bootstrap samples for all supported use cases"), too. I agree that it's simpler, but I didn't mention it because it's potentially a lot less efficient (e.g. if you are only interested in SE but do enough resampling to ensure that CIs are valid). But as you said, there is another motivation: to make sure you have done enough resampling for the results to be valid. Let's make that our initial focus, in which case `n_resamples='auto'` as you suggest makes a lot of sense.

Here's the main loop:

```python3
    theta_hat_b = []

    batch_nominal = batch or n_resamples

    for k in range(0, n_resamples, batch_nominal):
        batch_actual = min(batch_nominal, n_resamples-k)
        # Generate resamples
        resampled_data = []
        for sample in data:
            resample = _bootstrap_resample(sample, n_resamples=batch_actual,
                                           random_state=random_state)
            resampled_data.append(resample)

        # Compute bootstrap distribution of statistic
        theta_hat_b.append(statistic(*resampled_data, axis=-1))
    theta_hat_b = np.concatenate(theta_hat_b, axis=-1)
```

`theta_hat_b` holds the bootstrap distribution - the values of the statistic calculated for all the resamples so far. For simplicity of implementation, it is currently a list of arrays, each containing `batch_actual` values of the statistic, but ideally it would have been a pre-allocated array of length `n_resamples` (along the last axis), in which case `k` would serve as an index to drop each new batch of values into the right spot.

One way to implement this enhancement would be to change four lines.
* `theta_hat_b = []` would change to e.g. `bootman = bootstrap_manager(n_resamples, batch_nominal)`. Depending on the value of `n_resamples`, `bootstrap_manager` would return an object that will both store the bootstrap distribution and serve as an iterator for the loop (and either stop after `n_resamples` have been performed or, if `n_resamples` is a string, after the criteria for termination have been met).
* `for k in range(0, n_resamples, batch_nominal)` would become e.g. `for k in bootman`. The object `bootman` refers to is an [iterable](https://www.w3schools.com/python/python_iterators.asp).
* `theta_hat_b.append(statistic(*resampled_data, axis=-1))` would become e.g. `bootman.update(statistic(*resampled_data, axis=-1))`. Updating the bootstrap manager means giving it the new values of the statistic so it can decide whether to continue iterating or not.
* `theta_hat_b = np.concatenate(theta_hat_b, axis=-1)` would become e.g. `theta_hat_b = bootman.finish()`. The bootstrap manager is done and hands over a single array containing all the values of the statistic.

To match the existing behavior, the class might look like:
```
class BootstrapManager:
    def __init__(self, n_resamples, batch_nominal):
        self.theta_hat_b = []
        self.iter = range(0, n_resamples, batch_nominal)

    def __iter__(self):
        return iter(self.iter)

    def update(self, new_values):
        self.theta_hat_b.append(new_values)

    def finish(self):
        return np.concatenate(self.theta_hat_b, axis=-1)
```

I opened a PR on my own repo (mdhaber/scipy#60) to illustrate the idea. I could also add in the machinery for input validation and choosing which bootstrap manager class to instantiate based on the value of `n_resamples`; all you would need to do is implement such a class.
